
---

# **Pandas pour la Data Science**

## **Introduction aux DataFrames**

---

### **Introduction**

Le module `pandas` a été développé pour apporter à `Python` les outils nécessaires pour manipuler et analyser de gros volumes de données.

`Pandas` introduit la classe **`DataFrame`**, une structure de données qui s'apparente à un tableau et qui propose une manipulation et exploration de données plus avancées que les arrays `NumPy`.

Les principales fonctionnalités de `pandas` sont :

- la récupération des données depuis des fichiers (CSV, tableaux Excel, etc.) .
- la manipulation de ces données (suppression/ajout, modification, visualisation statistique, etc.).

Ce notebook a pour objectif de :

- Comprendre le format d'un `DataFrame`.
- Créer un premier `Dataframe`.
- Réaliser une première exploration d'un dataset (jeu de données en français) grâce à la classe `DataFrame`.

- **(a)** Importer le module `pandas` sous le nom `pd`.


In [1]:
# A
import pandas as pd

## **1. Format d'un DataFrame**

Un `DataFrame` se présente sous forme d'une **matrice** dont chaque ligne et chaque colonne porte un **indice**.
En général, les colonnes sont indexées par leur nom.

Un `DataFrame` sert à stocker des **bases de données**.
Les différentes **entrées** de la base (individus, animaux, objets, etc.) sont les différentes **lignes** et leurs **caractéristiques** sont les différentes **colonnes** :

|       | Nom   | Sexe | Taille | Age |
|-------|-------|------|--------|-----|
|0| Robert| M    | 174    | 23  |
|1| Mark  | M    | 182    | 40  |
|2| Aline | F    | 169    | 56  |

- Le `DataFrame` ci-dessus regroupe des informations sur **3 individus** : le tableau possède donc **3 lignes**.
- Pour chacun de ces individus, nous disposons de **4 variables** (le nom, le sexe, la taille et l'âge) : le tableau possède donc **4 colonnes**.

La colonne contenant **les numérotations des lignes** est appelée **l'index** et ne se gère pas de la même façon qu'une colonne du dataset.

On peut laisser l'index par défaut (numérotation des lignes), indexer avec une des colonnes du `DataFrame` ou indexer avec une liste que l'on définit nous-même.

Exemple : Indexation par défaut (numérotation des lignes) :

|       | Nom   | Sexe | Taille | Age |
|-------|-------|------|--------|-----|
|0| Robert| M    | 174    | 23  |
|1| Mark  | M    | 182    | 40  |
|2| Aline | F    | 169    | 56  |

Exemple : Indexation par la colonne `'Nom'` :

| Nom    | Sexe | Taille | Âge |
|--------|------|--------|-----|
| Robert | M    | 174    | 23  |
| Mark   | M    | 182    | 40  |
| Aline  | F    | 169    | 56  |

Exemple : Indexation par la liste `['personne_1', 'personne_2', 'personne_3']` :

| Nom        | Sexe | Taille | Âge |
|------------|------|--------|-----|
| personne_1 | Robert | M    | 174  | 23  |
| personne_2 | Mark   | M    | 182  | 40  |
| personne_3 | Aline  | F    | 169  | 56  |

Nous détaillerons plus loin comment définir l'index lors de la création d'un `DataFrame`.

La classe `DataFrame` présente plusieurs avantages par rapport à un array `Numpy` :

- Visuellement, un `DataFrame` est beaucoup plus **lisible** grâce à une indexation des colonnes et des lignes plus explicite.
- Au sein d'une même colonne les éléments sont du même type mais d'une colonne à l'autre, le **type des éléments peut varier*, ce qui n'est pas le cas des arrays `Numpy` qui ne supportent que des données de même type.
- La classe `DataFrame` contient davantage de méthodes pour la manipulation et le pré-traitement de bases de données, tandis que `NumPy` se spécialise plutôt dans le calcul optimisé.

 ## **2. Création d'un DataFrame à partir d'un array NumPy**

 Il est possible de créer directement un `DataFrame` à partir d'un array `NumPy` grâce au constructeur `DataFrame()`.
L'inconvénient de cette méthode est qu'elle n'est pas très pratique et le type des données est obligatoirement le même pour toutes les colonnes.

Regardons d'un peu plus près l'en-tête de ce constructeur.

```python
 pd.DataFrame(data, index, columns, ...)
```
- Le paramètre `data` contient les **données** à mettre en forme (array `NumPy`, liste, dictionnaire ou un autre `DataFrame`).
- Le paramètre `index`, si précisé, doit être une **liste** contenant les **indices des entrées**.
- Le paramètre `columns`, si précisé, doit être une **liste** contenant le **nom des colonnes**.

>  Pour les autres paramètres, vous pouvez consulter la [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) Python.

Exemple :
```python
# Création d'un array NumPy
array = np.array([[1, 2, 3, 4],
                  [5, 6, 7, 8],
                  [9, 10, 11, 12]])

# Instanciation d'un DataFrame
df = pd.DataFrame(data = array,                 # Les données à mettre en forme
                  index = ['i_1', 'i_2', 'i_3'],  # Les indices de chaque entrée
                  columns = ['A', 'B', 'C', 'D']) # Le nom des colonnes
```

Ceci produit le `DataFrame` suivant :

|     | A  | B  | C  | D  |
|-----|----|----|----|----|
| i_1 | 1  | 2  | 3  | 4  |
| i_2 | 5  | 6  | 7  | 8  |
| i_3 | 9  | 10 | 11 | 12 |


## **3. Création d'un DataFrame à partir d'un dictionnaire**

Une autre méthode pour créer un `DataFrame` est d'utiliser un dictionnaire.
Grâce à cette technique, les colonnes peuvent être de type différent et sont déjà définies lors de la création du `DataFrame`.

Exemple :
```python
# Création d'un dictionnaire
dictionnaire = {'A': [1, 5, 9],
                'B': [2, 6, 10],
                'C': [3, 7, 11],
                'D': [4, 8, 12]}

# Instanciation d'un DataFrame
df = pd.DataFrame(data = dictionnaire,
                  index = ['i_1', 'i_2', 'i_3'])
```

Ceci produit le même `DataFrame` que précédemment :

|     | A  | B  | C  | D  |
|-----|----|----|----|----|
| i_1 | 1  | 2  | 3  | 4  |
| i_2 | 5  | 6  | 7  | 8  |
| i_3 | 9  | 10 | 11 | 12 |

Le directeur d'une épicerie recense les informations suivantes sur son stock de produits alimentaires :

1.  **100** pots de miel dont la date d'expiration est le **10/08/2025** et valant **2€** l'unité.
2.  **55** paquets de farine expirant le **25/09/2024** coûtant chacun **3€**.
3.  **1800** bouteilles de vin à **1

- **(a)** À partir d'un **dictionnaire**, créer et afficher le `DataFrame` **`df`** qui **pour chaque produit** doit contenir de manière organisée :
  - Son nom.
  - Sa date d'expiration.
  - Sa quantité.
  - Son prix à l'unité.

Vous choisirez des noms de colonne pertinents et l'index sera celui par défaut (dans ce cas on ne spécifie pas le paramètre `index`).

In [10]:
# A
stock = [
    {
        'Nom': 'Miel',
        'Date d\'expiration': '10/08/2025',
        'Quantité': 100,
        'Prix à l\'unité': 2
    },
    {
        'Nom': 'Farine',
        'Date d\'expiration': '25/09/2024',
        'Quantité': 55,
        'Prix à l\'unité': 3
    },
    {
        'Nom': 'Vin',
        'Date d\'expiration': 'N/A',
        'Quantité': 1800,
        'Prix à l\'unité': 1
    }]

df = pd.DataFrame(data = stock)

print(df)

      Nom Date d'expiration  Quantité  Prix à l'unité
0    Miel        10/08/2025       100               2
1  Farine        25/09/2024        55               3
2     Vin               N/A      1800               1


 ## **4. Création d'un DataFrame à partir d'un fichier de données**

 Le plus souvent, les `DataFrame` sont directement créés à partir de fichiers contenant les données d'intérêt.
Cela peut être un fichier de format CSV, Excel, Texte etc...

Le format le plus courant est le format CSV, qui signifie *Comma-Separated Values* et désigne un fichier de type tableur dont les valeurs sont séparées par des virgules.

En voici un exemple :
```python
A, B, C, D,
1, 2, 3, 4,
5, 6, 7, 8,
9, 10, 11, 12
```

Dans ce format :

- **La première ligne contient le nom des colonnes**, mais il arrive que le nom des colonnes **ne soit pas renseigné**.
- Chaque **ligne** correspond à une **entrée** de la base de données.
- Les valeurs sont séparées par **un caractère de séparation**. Dans cet exemple, il s'agit de `','` mais cela pourrait être un `';'`.

Pour importer ces données dans un `DataFrame`, on utilise alors la fonction `read_csv` de `pandas` dont l'en-tête est la suivante :
```python
pd.read_csv(filepath_or_buffer , sep = ',', header = 0, index_col = 0 ... )
```
Les **arguments essentiels** de la fonction pd.read_csv à connaître sont :

>- `filepath_or_buffer` : Le **chemin d'accès du fichier** .csv relativement à l'environnement d'exécution. Si le fichier se trouve dans le même dossier que l'environnement Python, il suffit de renseigner le nom du fichier. Ce chemin doit être renseigné sous forme de **chaîne de caractères**.
>- `sep` : Le caractère utilisé dans le fichier .csv pour **séparer** les différentes colonnes. Cet argument doit être spécifié sous forme de **caractère**.
>- `header` : **Le numéro de la ligne qui contient les noms des colonnes**. Si par exemple les noms de colonnes sont renseignés dans la première ligne du fichier `.csv`, alors il faut spécifier **`header = 0`**. Si les noms ne sont pas renseignés, on mettra `header = None`.
>- `index_col` : Le **nom ou numéro de la colonne** contenant les indices de la base de données. Si les entrées de la base sont indexées par la première colonne, il faudra renseigner **`index_col = 0`**. Alternativement, si les entrées sont indexées par une colonne qui porte le nom `"Id"`, on pourra spécifier **`index_col = "Id"`**.

Cette fonction retournera un objet de type `DataFrame` qui contient toutes les données du fichier.

- **(a)** Charger les données contenues dans le fichier **`Transactions.csv`** dans un `DataFrame` nommé **`transactions`** :
> - Le fichier se trouve dans le **même dossier** que l'environnement de ce notebook.
>- Les colonnes sont séparées par une **virgule**.
>- Les noms des colonnes sont renseignés sur la **première ligne** du fichier.
>- Les lignes de la base sont indexées par la colonne **"transaction_id"** qui est aussi la **première colonne**.

In [17]:
# A
transactions = pd.read_csv('Transactions.csv', sep = ',', header = 0, index_col = 0)

transactions.head(20)

cust_id   tran_date  prod_subcat_code  prod_cat_code  Qty  \
transaction_id                                                              
80712190438      270351  28-02-2014                 1              1   -5   
29258453508      270384  27-02-2014                 5              3   -5   
51750724947      273420  24-02-2014                 6              5   -2   
93274880719      271509  24-02-2014                11              6   -3   
51750724947      273420  23-02-2014                 6              5   -2   
97439039119      272357  23-02-2014                 8              3   -2   
45649838090      273667  22-02-2014                11              6   -1   
22643667930      271489  22-02-2014                12              6   -1   
79792372943      275108  22-02-2014                 3              1   -3   
50076728598      269014  21-02-2014                 8              3   -4   
29258453508      270384  20-02-2014                 5              3    5   
25455265351      267750  20-02-2014                12              6    3   
1571002198       275023  20-02-2014                 6              5    4   
43134751727      268487  20-02-2014                 3              2   -1   
36554696014      269345  20-02-2014                 3              5    3   
56814940239      268799  20-02-2014                 7              5    5   
54295803788      270787  20-02-2014                12              5    5   
25963520987      274829  20-02-2014                 4              4    3   
17183929085      266863  20-02-2014                 1              2    1   
44783317894      269452  20-02-2014                 3              1    3   

                Rate      Tax  total_amt      Store_type  
transaction_id                                            
80712190438     -772  405.300  -4265.300          e-Shop  
29258453508    -1497  785.925  -8270.925          e-Shop  
51750724947     -791  166.110  -1748.110        TeleShop  
93274880719    -1363  429.345  -4518.345          e-Shop  
51750724947     -791  166.110  -1748.110        TeleShop  
97439039119     -824  173.040  -1821.040        TeleShop  
45649838090    -1450  152.250  -1602.250          e-Shop  
22643667930    -1225  128.625  -1353.625        TeleShop  
79792372943     -908  286.020  -3010.020             MBR  
50076728598     -581  244.020  -2568.020          e-Shop  
29258453508     1497  785.925   8270.925          e-Shop  
25455265351     1360  428.400   4508.400          e-Shop  
1571002198       587  246.540   2594.540          e-Shop  
43134751727     -611   64.155   -675.155          e-Shop  
36554696014     1253  394.695   4153.695          e-Shop  
56814940239      368  193.200   2033.200          e-Shop  
54295803788      584  306.600   3226.600          e-Shop  
25963520987      502  158.130   1664.130  Flagship store  
17183929085     1359  142.695   1501.695        TeleShop  
44783317894      825  259.875   2734.875        TeleShop

Nous avons chargé le fichier `Transactions.csv` dans le `DataFrame` **`transactions`** qui regroupe un historique de transactions effectuées entre 2011 et 2014.
Dans la section suivante, nous allons étudier ce dataset.

## **5. Première exploration d'un dataset grâce à la classe `DataFrame`**

La suite de ce notebook présente brièvement les principales **méthodes** de la classe `DataFrame` qui vont nous permettre de faire une rapide analyse de notre jeu de données, c'est-à-dire :

- Avoir un bref **aperçu des données** (méthode `head`, attributs `columns` et `shape`).
- Sélectionner des valeurs dans le `DataFrame` (méthodes `loc` et `iloc`).
- Réaliser une rapide **étude statistique** de nos données (méthodes `describe` et `value_counts`)

> Pour rappel, pour appliquer une méthode à un objet en Python (comme un DataFrame par exemple), il faut accoler la méthode en suffixe de l'objet. Exemple : mon_objet.ma_méthode()

## **6. Visualisation d'un `DataFrame`: méthode `head`, attributs `columns` et `shape`**

Il est possible d'avoir un aperçu d'un jeu de données en affichant **seulement les premières lignes** du `DataFrame`.

Pour cela, il faut utiliser la méthode **`head()`** en lui spécifiant en argument **le nombre de lignes** que nous souhaitons afficher (par défaut 5).

Il est aussi possible d'avoir un aperçu des **dernières lignes** en utilisant la méthode **`tail()`** qui s'applique de la même manière :
```python
# Affichage des 10 premières lignes d'un DataFrame mon_dataframe
mon_dataframe.head(10)
```
- **(a)** Afficher les **20 premières** lignes du `DataFrame` `transactions`.

In [14]:
# A
transactions.head(20)

cust_id   tran_date  prod_subcat_code  prod_cat_code  Qty  \
transaction_id                                                              
80712190438      270351  28-02-2014                 1              1   -5   
29258453508      270384  27-02-2014                 5              3   -5   
51750724947      273420  24-02-2014                 6              5   -2   
93274880719      271509  24-02-2014                11              6   -3   
51750724947      273420  23-02-2014                 6              5   -2   
97439039119      272357  23-02-2014                 8              3   -2   
45649838090      273667  22-02-2014                11              6   -1   
22643667930      271489  22-02-2014                12              6   -1   
79792372943      275108  22-02-2014                 3              1   -3   
50076728598      269014  21-02-2014                 8              3   -4   
29258453508      270384  20-02-2014                 5              3    5   
25455265351      267750  20-02-2014                12              6    3   
1571002198       275023  20-02-2014                 6              5    4   
43134751727      268487  20-02-2014                 3              2   -1   
36554696014      269345  20-02-2014                 3              5    3   
56814940239      268799  20-02-2014                 7              5    5   
54295803788      270787  20-02-2014                12              5    5   
25963520987      274829  20-02-2014                 4              4    3   
17183929085      266863  20-02-2014                 1              2    1   
44783317894      269452  20-02-2014                 3              1    3   

                Rate      Tax  total_amt      Store_type  
transaction_id                                            
80712190438     -772  405.300  -4265.300          e-Shop  
29258453508    -1497  785.925  -8270.925          e-Shop  
51750724947     -791  166.110  -1748.110        TeleShop  
93274880719    -1363  429.345  -4518.345          e-Shop  
51750724947     -791  166.110  -1748.110        TeleShop  
97439039119     -824  173.040  -1821.040        TeleShop  
45649838090    -1450  152.250  -1602.250          e-Shop  
22643667930    -1225  128.625  -1353.625        TeleShop  
79792372943     -908  286.020  -3010.020             MBR  
50076728598     -581  244.020  -2568.020          e-Shop  
29258453508     1497  785.925   8270.925          e-Shop  
25455265351     1360  428.400   4508.400          e-Shop  
1571002198       587  246.540   2594.540          e-Shop  
43134751727     -611   64.155   -675.155          e-Shop  
36554696014     1253  394.695   4153.695          e-Shop  
56814940239      368  193.200   2033.200          e-Shop  
54295803788      584  306.600   3226.600          e-Shop  
25963520987      502  158.130   1664.130  Flagship store  
17183929085     1359  142.695   1501.695        TeleShop  
44783317894      825  259.875   2734.875        TeleShop

- **(b)** Afficher les 10 **dernières** lignes du `DataFrame` `transactions`.

In [15]:
# B
transactions.tail(10)

cust_id   tran_date  prod_subcat_code  prod_cat_code  Qty  \
transaction_id                                                              
49882891062      271982  25-01-2011                10              5    4   
14787475597      273982  25-01-2011                 4              3    5   
50691119572      273031  25-01-2011                 6              5    1   
40893803228      272049  25-01-2011                11              6    3   
30856003613      266866  25-01-2011                 4              2    2   
94340757522      274550  25-01-2011                12              5    1   
89780862956      270022  25-01-2011                 4              1    1   
85115299378      271020  25-01-2011                 2              6    4   
72870271171      270911  25-01-2011                11              5    3   
77960931771      271961  25-01-2011                11              5    1   

                Rate      Tax  total_amt Store_type  
transaction_id                                       
49882891062     1330  558.600   5878.600     e-Shop  
14787475597      969  508.725   5353.725     e-Shop  
50691119572     1148  120.540   1268.540   TeleShop  
40893803228     1077  339.255   3570.255     e-Shop  
30856003613      444   93.240    981.240   TeleShop  
94340757522     1264  132.720   1396.720     e-Shop  
89780862956      677   71.085    748.085     e-Shop  
85115299378     1052  441.840   4649.840        MBR  
72870271171     1142  359.730   3785.730   TeleShop  
77960931771      447   46.935    493.935   TeleShop

Il est possible de récupérer le **nom des colonnes** d'un `DataFrame` grâce à son attribut `columns` :
```python
# Création d'un DataFrame df à partir d'un dictionnaire
dictionnaire = {'A': [1, 5, 9],
                'B': [2, 6, 10],
                'C': [3, 7, 11],
                'D': [4, 8, 12]}

df = pd.DataFrame(data = dictionnaire, index = ['i_1', 'i_2', 'i_3'])
```
Ces instructions produisent le même `DataFrame` que précédemment :

|     | A  | B  | C  | D  |
|-----|----|----|----|----|
| i_1 | 1  | 2  | 3  | 4  |
| i_2 | 5  | 6  | 7  | 8  |
| i_3 | 9  | 10 | 11 | 12 |

```python
# Affichage des colonnes du DataFrame df
 print(df.columns)
>>> ['A', 'B', 'C', 'D']
```
La liste du nom des colonnes est utile pour parcourir les colonnes d'un `DataFrame` à l'aide d'une boucle.

Il peut être intéressant de savoir combien de transactions (lignes) et combien de caractéristiques (colonnes) le dataset contient.

Pour cela nous allons utiliser l'attribut **`shape`** du `DataFrame` qui affiche les dimensions de notre `DataFrame` sous la forme d'un tuple (nombre de lignes, nombre de colonnes) :
```python
# Affichage des dimensions de mon_dataframe
print(mon_dataframe.shape)
>>> (3,4)
```

- **(c)** Afficher les dimensions du `DataFrame` `transactions` ainsi que **le nom de la 5ème colonne**. Rappelez-vous qu'en Python les indices commencent à 0.


In [18]:
# C
print(transactions.shape)
print(transactions.columns[4])

(23053, 9)
Qty


## **7. Sélection de colonnes d'un DataFrame**

L'extraction des colonnes d'un `DataFrame` est presque identique à l'extraction de données d'un dictionnaire.

Pour extraire une colonne d'un `DataFrame`, il suffit de renseigner **entre crochets** le **nom** de la colonne à extraire.
Pour extraire **plusieurs** colonnes, il faut mettre entre crochets **la liste des noms** des colonnes à extraire (donc 1 deuxième paire de crochet est nécessaire pour la liste) :
```python
# Affichage de la colonne 'cust_id'
print(transactions['cust_id'])

# Extraction des colonnes 'cust_id' et 'Qty' de transactions
cust_id_qty = transactions[["cust_id","Qty"]]
```

`cust_id_qty` est un **nouveau** `DataFrame` ne contenant que les colonnes `'cust_id'` et `'Qty'`.
L'affichage des 3 premières lignes de **`cust_id_qty`** donne :

| transactions_id | cust_id | Qty  |
|-----------------|---------|------|
| 80712190438     | 270351  | -5   |
| 29258453508     | 270384  | -5   |
| 51750724947     | 273420  | -2   |

Lorsque nous préparons un jeu de données pour l'exploiter plus tard, il est préférable de **séparer** les variables **catégorielles** des variables **quantitatives** :

- Une variable *catégorielle* est une variable qui contient des *modalités* ou *catégories* sans relation d'ordre. Par exemple, les variables **couleur préférée, pays et nationalité** sont des variables catégorielles
Les variables catégorielles du `DataFrame` `transactions` sont : `['cust_id', 'tran_date', 'prod_subcat_code', 'prod_cat_code', 'Store_type']`
- Une variable *quantitative* est une variable qui mesure une quantité, régie par une relation d'ordre permettant de comparer les éléments entre eux. Par exemple, les variables **taille, poids et âge** sont des variables quantitatives.
Les variables quantitatives de `transactions` sont : `['Qty', 'Rate', 'Tax', 'total_amt']`

Cette distinction est importante parce que certaines opérations basiques comme le calcul d'une moyenne n'a de sens que pour les variables quantitatives.

- **(a)** Dans un `DataFrame` nommé **`cat_vars`**, stocker les variables **catégorielles** de `transactions`.
- **(b)** Dans un `DataFrame` nommé **`num_vars`**, stocker les variables **quantitatives** de `transactions`.
- **(c)** Afficher les 5 premières lignes de chaque `DataFrame`.

In [19]:
# A
cat_vars = transactions[['cust_id', 'tran_date', 'prod_subcat_code', 'prod_cat_code', 'Store_type']]
num_vars = transactions[['Qty', 'Rate', 'Tax', 'total_amt']]

print(cat_vars.head(5))
print(num_vars.head(5))

                cust_id   tran_date  prod_subcat_code  prod_cat_code  \
transaction_id                                                         
80712190438      270351  28-02-2014                 1              1   
29258453508      270384  27-02-2014                 5              3   
51750724947      273420  24-02-2014                 6              5   
93274880719      271509  24-02-2014                11              6   
51750724947      273420  23-02-2014                 6              5   

               Store_type  
transaction_id             
80712190438        e-Shop  
29258453508        e-Shop  
51750724947      TeleShop  
93274880719        e-Shop  
51750724947      TeleShop  
                Qty  Rate      Tax  total_amt
transaction_id                               
80712190438      -5  -772  405.300  -4265.300
29258453508      -5 -1497  785.925  -8270.925
51750724947      -2  -791  166.110  -1748.110
93274880719      -3 -1363  429.345  -4518.345
51750724947      -2  -7

 ## **8. Sélection de lignes d'un DataFrame: méthodes loc et iloc**

 Pour extraire une ou plusieurs **lignes** d'un `DataFrame`, nous utilisons la méthode **`loc`**.
loc est un type de méthode très spécial car les arguments sont renseignés **entre crochets** et non entre parenthèses.
L'utilisation de cette méthode est très similaire à l'indexation des listes.

Afin de récupérer la ligne d'indice `i` d'un `DataFrame`, il suffit de renseigner `i` en argument de la méthode `loc` :
```python
# On récupère la ligne d'indice 80712190438 du DataFrame num_vars
print(num_vars.loc[80712190438])
>>>                 Rate    Tax  total_amt
>>> transaction_id                         
>>> 80712190438    -772.0  405.3    -4265.3
>>> 80712190438     772.0  405.3     4265.3
```
Afin de récupérer **plusieurs lignes**, nous pouvons soit :
- Renseigner une **liste d'indices**.
- Utiliser le **slicing** en précisant les indices de début et de fin de la plage.
```python
# On récupère les lignes d'indice 80712190438, 29258453508 et 51750724947 du DataFrame transactions
transactions.loc[[80712190438, 29258453508, 51750724947]]
```
`loc` peut aussi prendre en argument une colonne ou **liste de colonnes** afin d'affiner l'extraction de données :
```python
# On extrait les colonnes 'Tax' et 'total_amt' des lignes d'indices 80712190438 et 29258453508
transactions.loc[[80712190438, 29258453508], ['Tax', 'total_amt']]  
```
Cette instruction produit le `DataFrame` suivant :

| transaction_id | Tax     | total_amt  |
|----------------|---------|------------|
| 80712190438    | 405.300 | -4265.300  |
| 80712190438    | 405.300 | 4265.300   |
| 29258453508    | 785.925 | -8270.925  |
| 29258453508    | 785.925 | 8270.925   |


La méthode **`iloc`** permet d'indexer un `DataFrame` **exactement comme un array numpy**, c'est-à-dire en ne renseignant que les indexes numériques des lignes et colonnes. Ceci permet d'utiliser le slicing sans contraintes :
```python
# Extraction des 4 premières lignes et des 3 premières colonnes de transactions
transactions.iloc[0:4, 0:3]
```

Cette instruction produit le `DataFrame` suivant :

| transaction_id | cust_id | tran_date   | prod_subcat_code |
|----------------|---------|-------------|------------------|
| 80712190438    | 270351  | 28-02-2014  | 1.0              |
| 29258453508    | 270384  | 27-02-2014  | 5.0              |
| 51750724947    | 273420  | 24-02-2014  | 6.0              |
| 93274880719    | 271509  | 24-02-2014  | 11.0             |

Dans le cas où l'indexation des lignes est celle par défaut (numérotation des lignes), les méthodes loc et iloc sont **équivalentes**.

## **9. Indexation Conditionnelle d'un `DataFrame`**

Comme pour les arrays Numpy, nous pouvons utiliser l'**indexation conditionnelle** pour extraire les lignes d'un `Dataframe` qui vérifient une condition donnée.

Dans l'illustration suivante, nous sélectionnons les lignes du `DataFrame` `df` **pour lesquelles la colonne `col 2` vaut 3**.

![indexation_cond_final](https://github.com/diaBabPro/colabs/blob/main/indexation_cond_final.png?raw=true)

Il existe deux syntaxes pour indexer conditionnellement un `DataFrame` :
```python
# On sélectionne les lignes du DataFrame df pour lesquelles la colonne 'col 2' vaut 3.
df[df['col 2'] == 3]

df.loc[df['col 2'] == 3]
```

Si nous souhaitons **assigner** une nouvelle valeur à ces entrées, il faut absolument utiliser la méthode **`loc`**.
En effet, l'indexation avec la syntaxe `df[df['col 2'] == 3]` ne renvoie qu'une **copie** de ces entrées et ne permet pas d'accéder à l'emplacement mémoire où se trouvent les données.

Le gérant des transactions répertoriées dans le `DataFrame` **`transactions`** souhaite avoir accès aux **identifiants** des clients ayant fait un achat **en ligne** (c'est-à-dire dans un `"e-Shop"`) ainsi que **la date de la transaction correspondante**.

Nous avons les informations suivantes concernant les colonnes de `transactions` :

| Nom de la colonne | Description                                 |
|-------------------|---------------------------------------------|
| 'cust_id'         | Les identifiants des clients                |
| 'Store_type'      | Le type de magasin où s'est faite la transaction |
| 'tran_date'       | La date des transactions                    |

- **(a)** Dans un `DataFrame` nommé **`transactions_eshop`**, stocker les transactions qui ont lieu dans un magasin de type `"e-Shop"`.
- **(b)** Dans un autre `DataFrame` nommé **`transactions_id_date`**, stocker les identifiants des clients et la date des transactions du `DataFrame` `transactions_eshop`.
- **(c)** Afficher les 5 premières lignes de `transactions_id_date`.

In [20]:
# A
transactions_eshop = transactions[transactions['Store_type'] == 'e-Shop']

# B
transactions_id_date = transactions_eshop[['cust_id', 'tran_date']]

# C
print(transactions_id_date.head(5))

                cust_id   tran_date
transaction_id                     
80712190438      270351  28-02-2014
29258453508      270384  27-02-2014
93274880719      271509  24-02-2014
45649838090      273667  22-02-2014
50076728598      269014  21-02-2014


> À présent, le gérant voudrait avoir accès aux transactions effectuées par le client d'identifiant `268819`.

- **(d)** Dans un `DataFrame` nommé **`transactions_client_268819`**, stocker toutes les transactions dont l'identifiant du client est `268819`.
- **(e)** Une colonne d'un `DataFrame` peut être parcourue comme une liste dans une boucle `for`. À l'aide d'une boucle `for` sur la colonne `'total_amt'`, calculer et afficher le montant total des transactions du client `268819`.

In [23]:
# D
transactions_client_268819 = transactions[transactions['cust_id'] == 268819]

# E
total = 0
for i in transactions_client_268819['total_amt']:
  total += i

# Afficher le total arrondi à 2 décimales
print(f"{total:.2f}")

14911.97


## **10. Rapide étude statistique des données d'un `DataFrame`**

La méthode **`describe`** d'un `DataFrame` retourne un résumé des statistiques descriptives (min, max, moyenne, quantiles,..) de ses variables **quantitatives**.
C'est donc un outil très utile pour une première visualisation du type et de la distribution de ces variables.

Pour analyser les variables **catégorielles**, il est préférable de commencer par utiliser la méthode **`value_counts`** qui renvoie le nombre d'occurrences pour chaque modalité de ces variables. La méthode `value_counts` ne peut pas s'utiliser directement sur un `DataFrame` mais que sur les colonnes du `DataFrame` qui sont des objets de la classe **`pd.Series`**.

- **(a)** Utiliser la méthode `describe` du `DataFrame` `transactions`.
- **(b)** Les variables quantitatives sont `'Qty'`, `'Rate'`, `'Tax'` et `total_amt`. Est-ce que par défaut les statistiques produites par la méthode `describe` ne sont calculées **que** sur les variables quantitatives ?
- **(c)** Afficher le nombre d'occurrences de chaque modalité que prend la variable `Store_type` à l'aide de la méthode `value_counts`.

In [25]:
# A
transactions.describe()

# B
# Non

# C
print(transactions['Store_type'].value_counts())

Store_type
e-Shop            9311
MBR               4661
Flagship store    4577
TeleShop          4504
Name: count, dtype: int64


La méthode `describe` a calculé des statistiques sur les variables `cust_id`, `prod_subcat_code` et `prod_cat_code` alors que celles-ci sont des variables **catégorielles**.

Bien sûr, ces statistiques n'ont **aucun sens**. La méthode `describe` a traité ces variables comme quantitatives car les modalités qu'elles prennent sont de type numérique. C'est pourquoi il faut **faire attention** aux résultats retournés par la méthode `describe` et toujours **prendre du recul** sur ce que représentent les variables contenues dans le `DataFrame`.

Le gérant des transactions souhaite faire un rapport rapide sur les caractéristiques des transactions : il souhaite notamment connaître le **montant moyen dépensé** ainsi que la **quantité maximale** achetée.

- **(d)** Quel est le montant total moyen dépensé ? On s'intéressera à la colonne `'total_amt'` de **`transactions`**.
- **(e)** Quelle est la quantité maximale achetée ? On s'intéressera à la colonne `'Qty'` de **transactions**.

In [29]:
# D
print((transactions['total_amt'].mean()).round(2))

# E
print(transactions['Qty'].max())

2107.31
5


Certaines transactions ont des montants **négatifs**.
Il s'agit de transactions qui ont été annulées et remboursées au client.
Ces montants vont perturber la distribution des montants ce qui nous donne de **mauvaises estimations** des moyennes et quantiles de la variable `total_amt`.

- **(f)** Quelle est la moyenne du montant des transactions dont le montant est **positif** ?

In [30]:
# F
print((transactions[transactions['total_amt'] > 0]['total_amt'].mean()).round(2))

2608.44


# **Conclusion et récap**

La classe `DataFrame` du module `pandas` sera votre structure de données de choix pour explorer, analyser et traiter des bases de données.

Dans cette brève introduction, vous avez appris à :

- Créer un `DataFrame` à partir d'un array `numpy` et d'un dictionnaire à l'aide du constructeur **`pd.DataFrame`**.
- Créer un `DataFrame` à partir d'un fichier .csv à l'aide de la fonction **`pd.read_csv`**.
- Visualiser les premières et dernières lignes d'un `DataFrame` à l'aide des méthodes **`head`** et **`tail`**.
- Sélectionner une ou plusieurs colonnes d'un **`DataFrame`** en renseignant leurs noms entre crochets comme pour un dictionnaire.
- Sélectionner une ou plusieurs lignes d'un `DataFrame` en renseignant leur indice à l'aide des méthodes **`loc`** et **`iloc`**.
- Sélectionner les lignes d'un `DataFrame` qui vérifient une condition spécifique à l'aide de l'**indexation conditionnelle**.
- Effectuer une rapide étude statistique des variables quantitatives d'un `DataFrame` à l'aide de la méthode **`describe`**.

Le dataset `transactions` que nous avons utilisé est très propre : les variables sont proprement renseignées et ne contiennent pas de valeurs manquantes.
En pratique, ceci est **rarement** le cas, c'est pourquoi dans la suite nous verrons comment nettoyer des datasets à l'aide de `pandas`.